cluster/images/etcd/migrate/migrate_client.go

```diff
@@ -0,0 +1,223 @@
+		if err != nil {
+			return err
+		}
+		_, err = v2client.Set(context.Background(), key, value, nil)
+		return err
+	}
+	v3client, err := e.clientV3()
+	if err != nil {
+		return err
+	}
+	defer v3client.Close()
+	_, err = v3client.KV.Put(context.Background(), key, value)
+	return err
+}
+
+// Get reads a single value for a given key.
+func (e *CombinedEtcdClient) Get(version *EtcdVersion, key string) (string, error) {
+	if version.major == 2 {
+		v2client, err := e.clientV2()
+		if err != nil {
+			return "", err
+		}
+		resp, err := v2client.Get(context.Background(), key, nil)
+		if err != nil {
+			return "", err
+		}
+		return resp.Node.Value, nil
+	}
+	v3client, err := e.clientV3()
+	if err != nil {
+		return "", err
+	}
+	defer v3client.Close()
+	resp, err := v3client.KV.Get(context.Background(), key)
+	if err != nil {
+		return "", err
+	}
+	kvs := resp.Kvs
+	if len(kvs) != 1 {
+		return "", fmt.Errorf("expected exactly one value for key %s but got %d", key, len(kvs))
+	}
+
+	return string(kvs[0].Value), nil
+}
+
+func (e *CombinedEtcdClient) clientV2() (clientv2.KeysAPI, error) {
+	v2client, err := clientv2.New(clientv2.Config{Endpoints: []string{e.endpoint()}})
+	if err != nil {
+		return nil, err
+	}
+	return clientv2.NewKeysAPI(v2client), nil
+}
+
+func (e *CombinedEtcdClient) clientV3() (*clientv3.Client, error) {
+	return clientv3.New(clientv3.Config{Endpoints: []string{e.endpoint()}})
+}
+
+// Backup creates a backup of an etcd2 data directory at the given backupDir.
+func (e *CombinedEtcdClient) Backup(version *EtcdVersion, backupDir string) error {
+	// We cannot use etcd/client (v2) to make this call. It is implemented in the etcdctl client code.
+	if version.major != 2 {
+		return fmt.Errorf("etcd 2.x required but got version '%s'", version)
+	}
+	return e.runEtcdctlCommand(version,
+		"--debug",
+		"backup",
+		fmt.Sprintf("--data-dir=%s", e.cfg.dataDirectory),
+		fmt.Sprintf("--backup-dir=%s", backupDir),
+	)
+}
+
+// Snapshot captures a snapshot from a running etcd3 server and saves it to the given snapshotFile.
+// We cannot use etcd/clientv3 to make this call. It is implemented in the etcdctl client code.
+func (e *CombinedEtcdClient) Snapshot(version *EtcdVersion, snapshotFile string) error {
+	if version.major != 3 {
+		return fmt.Errorf("etcd 3.x required but got version '%s'", version)
+	}
+	return e.runEtcdctlCommand(version,
+		"--endpoints", e.endpoint(),
+		"snapshot", "save", snapshotFile,
+	)
+}
+
+// Restore restores a given snapshotFile into the data directory specified this clients config.
+func (e *CombinedEtcdClient) Restore(version *EtcdVersion, snapshotFile string) error {
+	// We cannot use etcd/clientv3 to make this call. It is implemented in the etcdctl client code.
+	if version.major != 3 {
+		return fmt.Errorf("etcd 3.x required but got version '%s'", version)
+	}
+	return e.runEtcdctlCommand(version,
+		"snapshot", "restore", snapshotFile,
+		"--data-dir", e.cfg.dataDirectory,
```

Comment:
Would be nice to be consistent about --data-dir value and --data-dir=value (In Backup() above) Ditto all other flags.

Reply:
Yes, this is an artifact from how the flags were defined inconsistently in the bash script which I didn't clean up in the translation. I'll do that now.